app/models/spree/order_contents.rb

```diff
@@ -11,25 +11,79 @@ def initialize(order)
     # Get current line item for variant if exists
     # Add variant qty to line_item
     def add(variant, quantity = 1, shipment = nil)
-      line_item = order.find_line_item_by_variant(variant)
-      add_to_line_item(line_item, variant, quantity, shipment)
+      line_item = add_to_line_item(variant, quantity, shipment)
+      update_shipment(shipment)
+      update_order
+      line_item
     end
 
     # Get current line item for variant
     # Remove variant qty from line_item
-    def remove(variant, quantity = 1, shipment = nil)
-      line_item = order.find_line_item_by_variant(variant)
+    def remove(variant, quantity = nil, shipment = nil)
+      line_item = remove_from_line_item(variant, quantity, shipment)
+      update_shipment(shipment)
+      order.update_order_fees! if order.completed?
+      update_order
+      line_item
+    end
 
-      unless line_item
-        raise ActiveRecord::RecordNotFound, "Line item not found for variant #{variant.sku}"
+    def update_or_create(variant, attributes)
+      line_item = find_line_item_by_variant(variant)
+
+      if line_item
+        line_item.update(attributes)
+      else
+        line_item = Spree::LineItem.new(attributes)
+        line_item.variant = variant
+        line_item.price = variant.price
+        order.line_items << line_item
       end
 
-      remove_from_line_item(line_item, variant, quantity, shipment)
+      order.reload
+      line_item
+    end
+
+    def update_cart(params)
+      if order.update_attributes(params)
+        discard_empty_line_items
+        update_shipment
+        update_order
+        true
+      else
+        false
+      end
+    end
+
+    def update_item(line_item, params)
+      if line_item.update_attributes(params)
+        discard_empty_line_items
+        order.update_line_item_fees! line_item
+        order.update_order_fees! if order.completed?
+        update_shipment
+        update_order
+        true
+      else
+        false
+      end
```

Comment:
I think these methods taking params are clearly breaking the single responsibility principle, a class should do one thing. What if you create a separate class for this? Or even two classes: one to update order based on params, another to update a line item based on params... you will know better.

Reply:
So this is another bit pulled straight from Spree 2.2. It's moved here to keep this logic out of an admin and an API controller, but it's used where the params are a hash of line items data (with `accepts_nested_attributes`).

Reply:
So the params are basically:
```
{order:
  line_items_attributes: [
    id: 1, quantity: 5,
    id: 4, quantity: 8,
    id: 6, quantity: 9
  ]
}
```
and it's more of a bulk-update of the order's line items than an update of the order...

Reply:
So the responsibility here is still roughly "performing operations that change line items, including any updates required as part of those changes".